File: 1020_Number_of_Enclaves.py
class Solution:
    def numEnclaves(self, grid):
        m, n = len(grid), len(grid[0])
        
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        def dfs(i, j):
            if i < 0 or j < 0 or i >= m or j >= n:
                return 0, False
            if grid[i][j] == 0:
                return 0, True

            enclosed = True
            if i == 0 or j == 0 or i == m - 1 or j == n - 1:
                enclosed = False

            grid[i][j] = 0
            size = 1

            for k in range(4):
                s, e = dfs(i + dx[k], j + dy[k])
                size += s
                enclosed = enclosed and e

            return size, enclosed

        total_cells = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    size, is_enclosed = dfs(i, j)
                    if is_enclosed:
                        total_cells += size

        return total_cells

File: cpp/1020_Number_of_Enclaves.cpp
class Solution 
{
public:
    int dx[4] = {-1, 1, 0, 0};
    int dy[4] = {0, 0, -1, 1};

    bool dfs(int i, int j, int m, int n, int& sz, vector<vector<int>>& grid)
    {
        if (i < 0 || j < 0 || i == m || j == n) return false;
        if (grid[i][j] == 0) return false;

        bool check = true;

        // island is on edge
        if (grid[i][j] == 1 && (i == 0 || j == 0 || i == m - 1 || j == n - 1)) check = false;

        grid[i][j] = 0;

        for (int k = 0; k < 4; ++k)
            if (dfs(i + dx[k], j + dy[k], m, n, grid))
                ++sz;

        return check;
    }

    int numEnclaves(vector<vector<int>>& grid) 
    {
        int m = grid.size(), n = grid[0].size();        

        int islands = 0;
        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                if (grid[i][j] == 1)
                {
                    int sz = 0;
                    if (dfs(i, j, m, n, sz, grid))
                        islands += sz;
                }
            }
        }

        return islands;
    }
};
